File: app/src/main/java/com/asfoundation/wallet/util/TransferParser.java
package com.asfoundation.wallet.util;

import com.asfoundation.wallet.entity.TransactionBuilder;
import com.asfoundation.wallet.interact.FindDefaultWalletInteract;
import com.asfoundation.wallet.repository.TokenRepositoryType;
import io.reactivex.Single;
import java.io.UnsupportedEncodingException;
import java.math.BigDecimal;
import java.util.Arrays;
import org.kethereum.erc681.ERC681;
import org.kethereum.erc681.ERC681ExtensionFunKt;
import org.kethereum.erc681.ERC681ParserKt;
import org.spongycastle.util.encoders.Hex;

public class TransferParser {
  private final FindDefaultWalletInteract findDefaultWalletInteract;
  private final TokenRepositoryType tokenRepository;

  public TransferParser(FindDefaultWalletInteract findDefaultWalletInteract,
      TokenRepositoryType tokenRepository) {
    this.findDefaultWalletInteract = findDefaultWalletInteract;
    this.tokenRepository = tokenRepository;
  }

  public Single<TransactionBuilder> parse(String uri) {
    if (ERC681ExtensionFunKt.isEthereumURLString(uri)) {
      return Single.just(ERC681ParserKt.parseERC681(uri))
          .flatMap(erc681 -> {
            switch (getTransactionType(erc681)) {
              case APPC:
                return buildAppcTransaction(erc681);
              case TOKEN:
                return buildTokenTransaction(erc681);
              case ETH:
              default:
                return buildEthTransaction(erc681);
            }
          });
    }
    return Single.error(new RuntimeException("is not an supported URI"));
  }

  private Single<TransactionBuilder> buildEthTransaction(ERC681 payment) {
    TransactionBuilder transactionBuilder = new TransactionBuilder("ETH");
    transactionBuilder.toAddress(payment.getAddress());
    transactionBuilder.amount(getEtherTransferAmount(payment));
    return Single.just(transactionBuilder);
  }

  private Single<TransactionBuilder> buildTokenTransaction(ERC681 payment) {
    return findDefaultWalletInteract.find()
        .flatMap(wallet -> tokenRepository.fetchAll(wallet.address)
            .flatMapIterable(Arrays::asList)
            .filter(token -> token.tokenInfo.address.equalsIgnoreCase(payment.getAddress()))
            .toList())
        .flatMap(tokens -> {
          if (tokens.isEmpty()) {
            return Single.error(new UnknownTokenException());
          } else {
            return Single.just(tokens.get(0));
          }
        })
        .map(token -> new TransactionBuilder(token.tokenInfo.symbol, token.tokenInfo.address,
            payment.getChainId(), getReceiverAddress(payment),
            getTokenTransferAmount(payment, token.tokenInfo.decimals), null,
            token.tokenInfo.decimals).shouldSendToken(true));
  }

  private Single<TransactionBuilder> buildAppcTransaction(ERC681 payment) {
    return findDefaultWalletInteract.find()
        .flatMap(wallet -> tokenRepository.fetchAll(wallet.address)
            .flatMapIterable(Arrays::asList)
            .filter(token -> token.tokenInfo.address.equalsIgnoreCase(payment.getAddress()))
            .toList())
        .flatMap(tokens -> {
          if (tokens.isEmpty()) {
            return Single.error(new UnknownTokenException());
          } else {
            return Single.just(tokens.get(0));
          }
        })
        .map(token -> new TransactionBuilder(token.tokenInfo.symbol, getIabContractAddress(payment),
            payment.getChainId(), getReceiverAddress(payment),
            getTokenTransferAmount(payment, token.tokenInfo.decimals), getSkuId(payment),
            token.tokenInfo.decimals));
  }

  private String getIabContractAddress(ERC681 payment) {
    return payment.getAddress();
  }

  private TransactionType getTransactionType(ERC681 payment) {
    if (payment.getFunction() != null && payment.getFunction()
        .equalsIgnoreCase("buy")) {
      return TransactionType.APPC;
    } else if (payment.getFunction() != null && payment.getFunction()
        .equalsIgnoreCase("transfer")) {
      return TransactionType.TOKEN;
    } else {
      return TransactionType.ETH;
    }
  }

  private String getSkuId(ERC681 payment) throws UnsupportedEncodingException {
    return new String(Hex.decode(payment.getFunctionParams()
        .get("data")
        .substring(2)
        .getBytes("UTF-8")));
  }

  private BigDecimal getEtherTransferAmount(ERC681 payment) {
    return convertToMainMetric(new BigDecimal(payment.getValue()), 18);
  }

  private BigDecimal getTokenTransferAmount(ERC681 payment, int decimals) {
    if (payment.getFunctionParams()
        .containsKey("uint256")) {
      return convertToMainMetric(new BigDecimal(payment.getFunctionParams()
          .get("uint256")), decimals);
    }
    return BigDecimal.ZERO;
  }

  private BigDecimal convertToMainMetric(BigDecimal value, int decimals) {
    try {
      StringBuilder divider = new StringBuilder(18);
      divider.append("1");
      for (int i = 0; i < decimals; i++) {
        divider.append("0");
      }
      return value.divide(new BigDecimal(divider.toString()));
    } catch (NumberFormatException ex) {
      return BigDecimal.ZERO;
    }
  }

  private String getReceiverAddress(ERC681 payment) {
    String address;
    if (payment.getFunction()
        .equals("transfer") || payment.getFunction()
        .equals("buy")) {
      address = payment.getFunctionParams()
          .get("address");
    } else {
      address = payment.getAddress();
    }
    return address;
  }

  public enum TransactionType {
    APPC, TOKEN, ETH
  }
}

File: app/src/main/java/com/asfoundation/wallet/ui/_TransferParser.kt
package com.asfoundation.wallet.ui

import com.appcoins.wallet.bdsbilling.repository.BdsApiSecondary
import com.appcoins.wallet.billing.repository.entity.TransactionData
import com.asf.appcoins.sdk.contractproxy.AppCoinsAddressProxySdk
import com.asf.wallet.BuildConfig
import com.asfoundation.wallet.entity.TransactionBuilder
import com.asfoundation.wallet.util.OneStepUri
import com.asfoundation.wallet.util.Parameters
import com.asfoundation.wallet.util.UnknownTokenException
import com.google.gson.Gson
import org.bouncycastle.util.encoders.Hex
import org.kethereum.erc681.ERC681
import java.math.BigDecimal
import java.math.BigInteger
import java.math.RoundingMode
import java.nio.charset.StandardCharsets
import javax.inject.Inject

class _EIPTransactionParser @Inject constructor(defaultTokenRepository: _DefaultTokenRepository) {

  private val tokenInfo = defaultTokenRepository.tokenInfo

  fun buildTransaction(erc681: ERC681): TransactionBuilder =
    when (erc681.function?.lowercase()) {
      "buy" -> buildAppcTransaction(erc681.checkToken())
      "transfer" -> buildTokenTransaction(erc681.checkToken())
      else -> buildEthTransaction(erc681)
    }

  private fun buildEthTransaction(payment: ERC681): TransactionBuilder =
    TransactionBuilder("ETH")
      .toAddress(payment.address)
      .amount(payment.etherTransferAmount)
      .apply { type = "inapp" } //Transfer only so it doesn't matter

  private fun buildTokenTransaction(payment: ERC681): TransactionBuilder =
    TransactionBuilder(
      tokenInfo.symbol,
      tokenInfo.address,
      payment.chainId,
      payment.receiverAddress,
      payment.getTokenTransferAmount(tokenInfo.decimals),
      tokenInfo.decimals
    ).shouldSendToken(true)

  private fun buildAppcTransaction(payment: ERC681): TransactionBuilder =
    payment.functionParams["data"]!!
      .substring(2)
      .toByteArray(StandardCharsets.UTF_8)
      .let { Hex.decode(it) }
      .toString(Charsets.UTF_8)
      .let { Gson().fromJson(it, TransactionData::class.java) }
      .let {
        TransactionBuilder(
          tokenInfo.symbol,
          payment.address,
          payment.chainId,
          payment.receiverAddress,
          payment.getTokenTransferAmount(tokenInfo.decimals),
          it.skuId,
          tokenInfo.decimals,
          payment.functionParams["iabContractAddress"],
          it.type,
          it.origin,
          it.domain,
          it.payload,
          null,
          it.orderReference,
          null,
          null,
          it.period,
          it.trialPeriod
        ).shouldSendToken(true)
      }

  private val ERC681.etherTransferAmount: BigDecimal
    get() = value?.toBigDecimal()
      ?.divide(BigDecimal(BigInteger.ONE, -18), 18, RoundingMode.DOWN)
      ?: BigDecimal.ZERO

  private fun ERC681.getTokenTransferAmount(decimals: Int): BigDecimal =
    functionParams["uint256"]?.toBigDecimal()
      ?.divide(BigDecimal(BigInteger.ONE, -decimals), decimals, RoundingMode.DOWN)
      ?: BigDecimal.ZERO

  private val ERC681.receiverAddress: String?
    get() = when (function?.lowercase()) {
      "transfer", "buy" -> functionParams["address"]
      else -> address
    }

  private fun ERC681.checkToken() = apply {
    if (!tokenInfo.address.equals(address, ignoreCase = true)) {
      throw UnknownTokenException()
    }
  }
}

class _OneStepTransactionParser @Inject constructor(
  private val proxySdk: AppCoinsAddressProxySdk,
  private val bdsApiSecondary: BdsApiSecondary,
  defaultTokenRepository: _DefaultTokenRepository
) {

  private val tokenInfo = defaultTokenRepository.tokenInfo
  private val cache: MutableMap<String, TransactionBuilder> = mutableMapOf()

  fun buildTransaction(oneStepUri: OneStepUri, referrerUrl: String): TransactionBuilder =
    cache[oneStepUri.toString()] ?: run {
      val chainId = if (oneStepUri.host == BuildConfig.PAYMENT_HOST_ROPSTEN_NETWORK) {
        Parameters.NETWORK_ID_ROPSTEN
      } else {
        Parameters.NETWORK_ID_MAIN
      }
      val domain = oneStepUri.parameters[Parameters.DOMAIN]

      val walletAddress = domain
        ?.let {
          bdsApiSecondary.getWallet(it).blockingGet().data.address
        }
        ?: oneStepUri.parameters[Parameters.TO]
        ?: throw _MissingWalletException()

      val paymentType = if (oneStepUri.parameters.containsKey(Parameters.SKILLS)) {
        Parameters.ESKILLS
      } else {
        Parameters.PAYMENT_TYPE_INAPP_UNMANAGED
      }
      val currency = oneStepUri.parameters[Parameters.CURRENCY] ?: "APPC"
      val amount = oneStepUri.parameters[Parameters.VALUE] ?: throw _MissingAmountException()
      val product = oneStepUri.parameters[Parameters.PRODUCT]
      val payload = oneStepUri.parameters[Parameters.DATA]
      val callbackUrl = oneStepUri.parameters[Parameters.CALLBACK_URL]
      val orderReference = oneStepUri.parameters[Parameters.ORDER_REFERENCE]
      val productToken = oneStepUri.parameters[Parameters.PRODUCT_TOKEN]

      // To avoid spending time, during the parse, getting the appc value by calling the conversion of
      // fiat to appc, we are setting the amount to zero. Later when this value is zero we should make
      // the request to get the conversion and set it on the transaction builder.
      val appcAmount: BigDecimal? =
        if (currency.uppercase() == "APPC") BigDecimal(amount).setScale(18)
        else BigDecimal.ZERO


      val networkId = if (BuildConfig.DEBUG) NETWORK_ID_ROPSTEN else NETWORK_ID_MAIN
      val iabContract = proxySdk.getIabAddress(networkId).blockingGet()
      val tokenContract = proxySdk.getAppCoinsAddress(networkId).blockingGet()
      TransactionBuilder(
        tokenInfo.symbol,
        tokenContract,
        chainId,
        walletAddress,
        appcAmount,
        product,
        tokenInfo.decimals,
        iabContract,
        paymentType,
        null,
        domain,
        payload,
        callbackUrl,
        orderReference,
        productToken,
        amount,
        currency,
        referrerUrl,
        ""
      ).shouldSendToken(true)
    }.also {
      cache[oneStepUri.toString()] = it
    }

  companion object {
    private const val NETWORK_ID_ROPSTEN = 3
    private const val NETWORK_ID_MAIN = 1
  }
}
